<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getModelFromKEGG</title>
  <meta name="keywords" content="getModelFromKEGG">
  <meta name="description" content="getModelFromKEGG">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../../index.html">Home</a> &gt;  <a href="../index.html">external</a> &gt; <a href="index.html">kegg</a> &gt; getModelFromKEGG.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../../index.html"><img alt="<" border="0" src="../../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for external\kegg&nbsp;<img alt=">" border="0" src="../../right.png"></a></td></tr></table>-->

<h1>getModelFromKEGG
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="box"><strong>getModelFromKEGG</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="box"><strong>function [model, KOModel]=getModelFromKEGG(keggPath,keepUndefinedStoich,keepIncomplete,keepGeneral) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getModelFromKEGG
   Retrieves information stored in KEGG database and generates a model

   keggPath            if keggGenes.mat, keggMets.mat, keggPhylDist.mat
                       or keggRxns.mat is not in the RAVEN\external\kegg
                       directory, this function will attempt to read data
                       from a local FTP dump of the KEGG database.
                       keggPath is the path to the root of this database
   keepUndefinedStoich include reactions in the form n A &lt;=&gt; n+1 A. These
                       will be dealt with as two separate metabolites
                       (opt, default true)
   keepIncomplete      include reactions which have been labelled as
                       &quot;incomplete&quot;, &quot;erroneous&quot; or &quot;unclear&quot; (opt,
                       default true)
   keepGeneral         include reactions which have been labelled as
                       &quot;general reaction&quot;. These are reactions on the form
                       &quot;an aldehyde &lt;=&gt; an alcohol&quot;, and are therefore
                       unsuited for modelling purposes. Note that not all
                       reactions have this type of annotation, and the
                       script will therefore not be able to remove all
                       such reactions (opt, default false)

   model               a model structure generated from the database.
                       All reactions and the metabolites used in them
                       will be added
   KOModel             a model structure representing the KEGG Orthology
                       ids and their associated genes. The KO ids are
                       saved as reactions

   Usage: getModelFromKEGG(keggPath,keepUndefinedStoich,keepIncomplete,keepGeneral)

   Eduard Kerkhoven, 2017-02-28</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../../matlabicon.gif)">
<li><a href="getGenesFromKEGG.html" class="code" title="function model=getGenesFromKEGG(keggPath,koList)">getGenesFromKEGG</a>	getGenesFromKEGG</li><li><a href="getMetsFromKEGG.html" class="code" title="function model=getMetsFromKEGG(keggPath)">getMetsFromKEGG</a>	getMetsFromKEGG</li><li><a href="getRxnsFromKEGG.html" class="code" title="function model=getRxnsFromKEGG(keggPath,keepUndefinedStoich,keepIncomplete, keepGeneral)">getRxnsFromKEGG</a>	getRxnsFromKEGG</li></ul>
This function is called by:
<ul style="list-style-image:url(../../matlabicon.gif)">
<li><a href="getKEGGModelForOrganism.html" class="code" title="function model=getKEGGModelForOrganism(organismID,fastaFile,dataDir,outDir,keepUndefinedStoich,keepIncomplete,keepGeneral,cutOff,minScoreRatioG,minScoreRatioKO,maxPhylDist,nSequences,seqIdentity)">getKEGGModelForOrganism</a>	getKEGGModelForOrganism</li></ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function [model, KOModel]=getModelFromKEGG(keggPath,keepUndefinedStoich,keepIncomplete,keepGeneral)</a>
0002 <span class="comment">% getModelFromKEGG</span>
0003 <span class="comment">%   Retrieves information stored in KEGG database and generates a model</span>
0004 <span class="comment">%</span>
0005 <span class="comment">%   keggPath            if keggGenes.mat, keggMets.mat, keggPhylDist.mat</span>
0006 <span class="comment">%                       or keggRxns.mat is not in the RAVEN\external\kegg</span>
0007 <span class="comment">%                       directory, this function will attempt to read data</span>
0008 <span class="comment">%                       from a local FTP dump of the KEGG database.</span>
0009 <span class="comment">%                       keggPath is the path to the root of this database</span>
0010 <span class="comment">%   keepUndefinedStoich include reactions in the form n A &lt;=&gt; n+1 A. These</span>
0011 <span class="comment">%                       will be dealt with as two separate metabolites</span>
0012 <span class="comment">%                       (opt, default true)</span>
0013 <span class="comment">%   keepIncomplete      include reactions which have been labelled as</span>
0014 <span class="comment">%                       &quot;incomplete&quot;, &quot;erroneous&quot; or &quot;unclear&quot; (opt,</span>
0015 <span class="comment">%                       default true)</span>
0016 <span class="comment">%   keepGeneral         include reactions which have been labelled as</span>
0017 <span class="comment">%                       &quot;general reaction&quot;. These are reactions on the form</span>
0018 <span class="comment">%                       &quot;an aldehyde &lt;=&gt; an alcohol&quot;, and are therefore</span>
0019 <span class="comment">%                       unsuited for modelling purposes. Note that not all</span>
0020 <span class="comment">%                       reactions have this type of annotation, and the</span>
0021 <span class="comment">%                       script will therefore not be able to remove all</span>
0022 <span class="comment">%                       such reactions (opt, default false)</span>
0023 <span class="comment">%</span>
0024 <span class="comment">%   model               a model structure generated from the database.</span>
0025 <span class="comment">%                       All reactions and the metabolites used in them</span>
0026 <span class="comment">%                       will be added</span>
0027 <span class="comment">%   KOModel             a model structure representing the KEGG Orthology</span>
0028 <span class="comment">%                       ids and their associated genes. The KO ids are</span>
0029 <span class="comment">%                       saved as reactions</span>
0030 <span class="comment">%</span>
0031 <span class="comment">%   Usage: getModelFromKEGG(keggPath,keepUndefinedStoich,keepIncomplete,keepGeneral)</span>
0032 <span class="comment">%</span>
0033 <span class="comment">%   Eduard Kerkhoven, 2017-02-28</span>
0034 <span class="comment">%</span>
0035 
0036 <span class="keyword">if</span> nargin&lt;2
0037     keepUndefinedStoich=true;
0038 <span class="keyword">end</span>
0039 <span class="keyword">if</span> nargin&lt;3
0040     keepIncomplete=true;
0041 <span class="keyword">end</span>
0042 <span class="keyword">if</span> nargin&lt;4
0043     keepGeneral=true;
0044 <span class="keyword">end</span>
0045 
0046 <span class="comment">%First get all reactions</span>
0047 model=<a href="getRxnsFromKEGG.html" class="code" title="function model=getRxnsFromKEGG(keggPath,keepUndefinedStoich,keepIncomplete, keepGeneral)">getRxnsFromKEGG</a>(keggPath,keepUndefinedStoich,keepIncomplete,keepGeneral);
0048 fprintf(<span class="string">'KEGG reactions loaded\n'</span>);
0049 
0050 <span class="comment">%Get the KO ids that are associated with any of the reactions. They will be</span>
0051 <span class="comment">%used later on to create a rxn-gene matrix</span>
0052 KOs=cell(numel(model.rxns)*2,1); <span class="comment">%Make room for two KO ids per reaction</span>
0053 
0054 addToIndex=1;
0055 <span class="comment">%Add all KO ids that are used</span>
0056 <span class="keyword">for</span> i=1:numel(model.rxns)
0057    <span class="keyword">if</span> isstruct(model.rxnMiriams{i})
0058       <span class="keyword">for</span> j=1:numel(model.rxnMiriams{i}.name)
0059          <span class="keyword">if</span> strcmpi(<span class="string">'urn:miriam:kegg.ko'</span>,model.rxnMiriams{i}.name{j});
0060             <span class="comment">%Add the KO id</span>
0061             KOs(addToIndex)=model.rxnMiriams{i}.value(j);
0062             addToIndex=addToIndex+1;
0063          <span class="keyword">end</span>
0064       <span class="keyword">end</span>
0065    <span class="keyword">end</span>
0066 <span class="keyword">end</span>
0067 
0068 KOs=KOs(1:addToIndex-1);
0069 KOs=unique(KOs);
0070 
0071 <span class="comment">%Get all genes from any organism in KEGG that is associated with any of</span>
0072 <span class="comment">%the KOs</span>
0073 KOModel=<a href="getGenesFromKEGG.html" class="code" title="function model=getGenesFromKEGG(keggPath,koList)">getGenesFromKEGG</a>(keggPath,KOs);
0074 fprintf(<span class="string">'KEGG genes loaded\n'</span>);
0075 
0076 model.genes=KOModel.genes;
0077 
0078 <span class="comment">%It can be that there are KOs from the reactions that have no database</span>
0079 <span class="comment">%entry. These are (as far as I've seen) lumped versions of other KOs and</span>
0080 <span class="comment">%should be removed</span>
0081 KOsToRemove=setdiff(KOs, KOModel.rxns);
0082 
0083 <span class="comment">%Loop through all reactions and delete the KOs that weren't found</span>
0084 <span class="keyword">for</span> i=1:numel(model.rxns)
0085     <span class="keyword">if</span> isstruct(model.rxnMiriams{i})
0086         <span class="keyword">for</span> j=1:numel(model.rxnMiriams{i}.name)
0087             toDel=[];
0088             <span class="keyword">if</span> strcmp(model.rxnMiriams{i}.name{j},<span class="string">'urn:miriam:kegg.ko'</span>)
0089                 <span class="keyword">if</span> ismember(model.rxnMiriams{i}.value{j},KOsToRemove)
0090                    toDel=[toDel;j];
0091                 <span class="keyword">end</span>
0092             <span class="keyword">end</span>
0093         <span class="keyword">end</span>
0094         <span class="comment">%Delete the KOs</span>
0095         <span class="keyword">if</span> any(toDel)
0096            <span class="comment">%If all posts are deleted, then delete the whole structure</span>
0097            <span class="keyword">if</span> numel(toDel)==j
0098                 model.rxnMiriams{i}=[];
0099            <span class="keyword">else</span>
0100                 model.rxnMiriams{i}.name(toDel)=[];
0101                 model.rxnMiriams{i}.value(toDel)=[];
0102            <span class="keyword">end</span>
0103         <span class="keyword">end</span>
0104     <span class="keyword">end</span>
0105 <span class="keyword">end</span>
0106 
0107 <span class="comment">%Create the rxnGeneMat for the reactions. This is simply done by merging</span>
0108 <span class="comment">%the gene associations for all the involved KOs</span>
0109 r=zeros(10000000,1); <span class="comment">%Store the positions since it's slow to write to a sparse array in a loop</span>
0110 c=zeros(10000000,1);
0111 counter=1;
0112 <span class="keyword">for</span> i=1:numel(model.rxns)
0113     <span class="keyword">if</span> isstruct(model.rxnMiriams{i})
0114         I=strncmp(<span class="string">'urn:miriam:kegg.ko'</span>,model.rxnMiriams{i}.name,18);
0115         [J, K]=ismember(model.rxnMiriams{i}.value(I),KOModel.rxns);
0116         <span class="comment">%Find all gene indexes that correspond to any of these KOs</span>
0117         [~, L]=find(KOModel.rxnGeneMat(K(J),:));
0118         <span class="keyword">if</span> any(L)
0119             <span class="comment">%Allocate room for more elements if needed</span>
0120             <span class="keyword">if</span> counter+numel(L)-1&gt;=numel(r)
0121                 r=[r;zeros(numel(r),1)];
0122                 c=[c;zeros(numel(c),1)];
0123             <span class="keyword">end</span>
0124             r(counter:counter+numel(L)-1)=ones(numel(L),1)*i;
0125             c(counter:counter+numel(L)-1)=L(:);
0126             counter=counter+numel(L);
0127         <span class="keyword">end</span>
0128     <span class="keyword">end</span>
0129 <span class="keyword">end</span>
0130 
0131 model.rxnGeneMat=sparse(r(1:counter-1),c(1:counter-1),ones(counter-1,1));
0132 <span class="keyword">if</span> size(model.rxnGeneMat,1)~=numel(model.rxns) || size(model.rxnGeneMat,2)~=numel(KOModel.genes)
0133     model.rxnGeneMat(numel(model.rxns),numel(KOModel.genes))=0;
0134 <span class="keyword">end</span>
0135 
0136 <span class="comment">%Then get all metabolites</span>
0137 metModel=<a href="getMetsFromKEGG.html" class="code" title="function model=getMetsFromKEGG(keggPath)">getMetsFromKEGG</a>(keggPath);
0138 fprintf(<span class="string">'KEGG metabolites loaded\n'</span>);
0139 
0140 <span class="comment">%Add information about all metabolites to the model</span>
0141 [a, b]=ismember(model.mets,metModel.mets);
0142 a=find(a);
0143 b=b(a);
0144 
0145 <span class="keyword">if</span> ~isfield(model,<span class="string">'metNames'</span>)
0146    model.metNames=cell(numel(model.mets),1);
0147    model.metNames(:)={<span class="string">''</span>};
0148 <span class="keyword">end</span>
0149 model.metNames(a)=metModel.metNames(b);
0150 
0151 <span class="keyword">if</span> ~isfield(model,<span class="string">'metFormulas'</span>)
0152    model.metFormulas=cell(numel(model.mets),1);
0153    model.metFormulas(:)={<span class="string">''</span>};
0154 <span class="keyword">end</span>
0155 model.metFormulas(a)=metModel.metFormulas(b);
0156 
0157 <span class="keyword">if</span> ~isfield(model,<span class="string">'inchis'</span>)
0158    model.inchis=cell(numel(model.mets),1);
0159    model.inchis(:)={<span class="string">''</span>};
0160 <span class="keyword">end</span>
0161 model.inchis(a)=metModel.inchis(b);
0162 
0163 <span class="keyword">if</span> ~isfield(model,<span class="string">'metMiriams'</span>)
0164    model.metMiriams=cell(numel(model.mets),1);
0165 <span class="keyword">end</span>
0166 model.metMiriams(a)=metModel.metMiriams(b);
0167 
0168 <span class="comment">%Put all metabolites in one compartment called 's' (for system). This is</span>
0169 <span class="comment">%done just to be more compatible with the rest of the code</span>
0170 model.comps={<span class="string">'s'</span>};
0171 model.compNames={<span class="string">'System'</span>};
0172 model.compOutside={<span class="string">''</span>};
0173 model.metComps=ones(numel(model.mets),1);
0174 
0175 <span class="comment">%If reactions with undefined stoichiometry are kept, then the corresponding</span>
0176 <span class="comment">%metabolites will have ids such as &quot;(n+1) C000001&quot; and their names will be</span>
0177 <span class="comment">%empty. These ids are not valid SBML identifiers and are therefore replaced</span>
0178 <span class="comment">%with &quot;undefined1, undefined2...&quot;. The former ids are stored as the new</span>
0179 <span class="comment">%names</span>
0180 I=find(cellfun(@any,strfind(model.mets,<span class="string">'n'</span>)) | cellfun(@any,strfind(model.mets,<span class="string">'m'</span>)));
0181 model.metNames(I)=model.mets(I);
0182 repNums=1:numel(I);
0183 repIDs=strcat(<span class="string">'undefined_'</span>,cellfun(@num2str,num2cell(repNums(:)),<span class="string">'UniformOutput'</span>,false));
0184 model.mets(I)=repIDs;
0185 
0186 <span class="comment">%It could also be that the metabolite names are empty for some other</span>
0187 <span class="comment">%reason. In that case, use the ID instead</span>
0188 I=cellfun(@isempty,model.metNames);
0189 model.metNames(I)=model.mets(I);
0190 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Tue 28-Feb-2017 23:41:43 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>